Bootstrap

slippers opened the pull request:
### about

this is an integration between flask-blogging and flask-bootstrap.  i believe that with these changes other flask extensions and applications that contain bootstrap will be easier to integrate.  this will break current implementations out there.
### setup
- create a templates folder in your application
- copy the base.html from flask-blogging/templates/blogging/base.html
- modify your base.html to reference your custom bootstrap css file.  create a static folder in your application and place your css file there.  in this case tango.css.  
  `
  {% block styles -%}  
  {{super()}}  
  <link rel=stylesheet type=text/css href="{{ url_for('static', filename='tango.css') }}">  
  {% endblock %}  
  `
### more

flask-bootstrap is applied when flask-blogging is first established in the engine.py file.  the Bootstrap extension adds a blueprint to hang all the function and templates it has onto your app.  the flask-blogging templates all inherit from base.html.  because of the way the jinja templates work only the application's template is probed for the inherited templates.  that is why it needs to be in the application space. 
### template changes

some of the templates changed 
- removal of bootstrap 2 stuff
- added toolbar
- modified delete post prompt

gouthambs commented:
@slippers  This is very interesting. Let me digest the changes a little bit.

If someone is not using flask-bootstrap, are we going to start forcing them to do that? Can a choice between flask-bootstrap or not be made optional?

slippers commented:
yes, should make this optional to have flask-bootstrap but still support it.  also don't want to break current implementations.  i will go back and try again.

to support flask-bootstrap would take some changes to the templates so that it aligns with the current version of bootstrap and what blocks it is expecting to find.  

### differences with flask-bootstrap
* uses a blueprint thus the application templates folder
* needs blocks named a specific way  
block [head, styles, content, scripts].  
ie: main -> content, extrastyle -> style, extrajs -> scripts
* uses the latest version of bootstrap

gouthambs commented:
I think, template folder for blueprint is configurable. Can we have two template folder bases side by side and point to the one based on whether someone is using flask-bootstrap or not.

gouthambs commented:
Also what advantages of incorporating flask bootstrap should we highlight?

slippers commented:
flask-bootstrap advantages
* static vs dynamic resource resolution
* lots of macros like wtf, ie fix, google analytics
* formal blocks
* flask-nav support
* pagination support through sql-alchemy

slippers commented:
there is now no requirement for flask-bootstrap.  

the extension templates have been changed to comply with flask-bootstrap block tags.

the page.html template modal dialog was using deprecated code. updated. 

### integrate flask-bootstrap into flask_blogging
* create templates/blogging directory in your application
* copy the flask_blogging/templates/blogging/base_flask_bootstrap.html into your templates/blogging directory.
* rename to base.html
* install the flask_bootstrap extension
* import Bootstrap -> from flask_bootstrap import Bootstrap
* register with your flask app -> Bootstrap(app)
* restart the application

gouthambs commented on flask_blogging/templates/blogging/base.html:
```diff
@@ -1,15 +1,16 @@
-    {% block style %}
+    {% block styles %}
     <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css" rel="stylesheet"/>
+    <link href="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-markdown/2.8.0/css/bootstrap-markdown.min.css"
```
This css is used only by the editor. That is why I had it in editor alone. Would that be compatible with flask-bootstrap in place? If not we can leave it in here.

gouthambs commented:
Here is another idea. Would like to hear what you think. Instead of a workflow where the user has to copy the templates and rename the base_flask_bootstrap.html to base.html. How about we rename the `base.html` to `base_plain.html`. Let's create a new `base.html` that says:

```
{% if config.BLOGGING_FLASK_BOOTSTRAP %}
{% include base_flask_bootstrap.html %}
{% else %}
{% include base_plain.html %}
{% endif %}

```
This way the whole choice of using flask-bootstrap or not can be a configuration change. Based on the config variable, we can do the `import flask_bootstrap` and adding the blueprint to the app.

So the `base.html` can switch between old and new versions. Then lets document the changes to `base_plain.html` to achieve compatibility with flask bootstrap. I think there is a file in release notes where I document compatibility aspects.

slippers commented:
your idea was right on!  this should make it painless.

i would like to write a test for this.  

### changes
* added Flask-Bootstrap to the requirements
* modified the engine.py to conditionally load Bootstrap
* BLOGGING_FLASK_BOOTSTRAP default to False
* base.html loads either base_orig.html or base_flask_bootstrap.html depending on BLOGGING_FLASK_BOOTSTRAP
* added a navbar block so now both templates look the same

gouthambs commented:
:+1:  This PR is coming along really well. I can see that you have improved the design greatly. We should think about

- documenting the changes (new blocks etc) so people can transition from the old to new version
- have a blog post written up documenting the improvements brought in this release and spread it on reddit etc

slippers commented:
help: having an code indentation issue.  error just moves from over to under indentation.
`
flask_blogging/engine.py:166:17: E126 continuation line over-indented for hanging indent
`
added more bootstrap code to the engine.py file to allow for the automatic configuration of the CDNs.  there are flask_bootstrap configs to control what url is created, local, http, https.
here is an example
`
<link href="{{bootstrap_find_resource('css/bootstrap-markdown.css', cdn='bootstrap-markdown')}}" rel="stylesheet"/>
`

gouthambs commented:
If you have long strings you can do something like:

```
string = "hello " \
         "world!"
```
This is the recommended way of getting strings to stretch out more than a line. Can you see if that helps. If not I am not sure I understand the question yet.

You can perhaps help me get up to speed on all the changes here. I see that you are doing some auto configuration for CDN etc. If people are plugging this into a site that they have built using flask-bootstrap already, would all this be conflicting with a setup they might already have.

gouthambs commented on flask_blogging/engine.py:
```diff
@@ -154,3 +156,17 @@ def process_post(self, post, render=True):
     def get_user_name(cls, user):
         user_name = user.get_name() if hasattr(user, "get_name") else str(user)
         return user_name
+
```
This should perhaps be called something like `static_asset_loader` because you seem to be loading more than bootstrap. 

Also there was a reason why I had markdown and bootstrap markdown load on the editor view. That way you are not loading unnecessary static assets for all the other pages. Google will ding sites like this with unoptimised static assets being loaded.

slippers commented:
made the recommended changes.  still working on a test for this feature.  it is very simple with a change in configuration.

slippers commented:
the bootstrap test works.  had to pass in the configuration prior to setup calling the engine.

gouthambs commented on test/test_views.py:
```diff
@@ -31,11 +31,12 @@ def _create_storage(self):
         return BloggingEngine(self.app, self.storage)
 
-    def setUp(self):
+    def setUp(self, **config):
```
The UnitTest standard signature for `setUp` seems to be be just `setUp(self)` per [unittest docs](https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.setUp).  One workaround is to implement `other_config()` method which will just return an empty dict in the parent class but can be overridden in `TestViewsWithBootstrap` to inject the required parameters.

More than anything, I am concerned if this could have the wrong `setUp` to be called down the line.

slippers replied:
good idea.  i will change

On Sat, Jul 23, 2016 at 1:35 PM, Goutham Balaraman <notifications@github.com
> wrote:

> In test/test_views.py
> <https://github.com/gouthambs/Flask-Blogging/pull/73#discussion_r71977892>
> :
>
> > @@ -31,11 +31,12 @@ def _create_storage(self):
> >      def _create_blogging_engine(self):
> >          return BloggingEngine(self.app, self.storage)
> >
> > -    def setUp(self):
> > +    def setUp(self, **config):
>
> The UnitTest standard signature for setUp seems to be be just setUp(self)
> per unittest docs
> <https://docs.python.org/2.7/library/unittest.html#unittest.TestCase.setUp>.
> One workaround is to implement other_config() method which will just
> return an empty dict in the parent class but can be overridden in
> TestViewsWithBootstrap to inject the required parameters.
>
> More than anything, I am concerned if this could have the wrong setUp to
> be called down the line.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gouthambs/Flask-Blogging/pull/73/files/6bc186a6713454d4b0b05c486aab39970533d51c..e8257e28afc601d672b89a428cbd547a4e95de4a#r71977892>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABy59bmReS-xd-K25HgEgVUeYd24prDIks5qYnsXgaJpZM4JIOM->
> .
>

gouthambs commented on test/test_views.py:
```diff
@@ -31,11 +31,15 @@ def _create_storage(self):
+    def other_config(self):
+        pass
+
+    def setUp(self, **config):
```
You perhaps forgot to change the signature of this function back to `setUp(self)`.

slippers commented:
hello,  still working on integrating this version of flask-blogging into a site already using flask-bootstrap.  seeing some issues and thinking about how to document. 

learning to use vim with all the python plugins.  thanks for the patience with this pr.  as far as issues:)  

flask_bootstrap creates a extension and blueprint called 'bootstrap'.  can only call Bootstrap(app) once for a flask application.  a blueprint can only be registered once.  the example site and the tests didn't call Bootstrap(app) twice.

a solution was to have the bootstrap extension detected in the engine and to make the appropriate changes.  just have to call Bootstrap(app) before registering the BloggingEngine() in the app.

gouthambs commented:
This PR is a huge change from the original code base. So I would just wait for you to test it out. So take your time.

On a side note, I use PyCharm community edition (free) for python coding. It is quite good, and I have been very happy with it. You may like it if you are comfortable using an IDE. If you want something lighter, then VIM is decent.

slippers commented:
modified the static_asset_loader.  added new test cases for flask_bootstrap.

gouthambs commented:
@slippers Just pinging to see if you have been using the flask_bootstrap in this project. Do you find the usage stable enough?

I am still waiting to understand if this PR would have a wider use case for other users or not. Would help if more people would share the need though.

slippers commented:
sorry for the slow response.

documentation is needed.
* integrate the bootstrap extension
* configuration for switching between base templates
* how to override the base template
* how to override the CDNs established in engine.py

was looking at another flask extension 'flask-admin' that has bootstrap
involved.  they don't use flask-bootstrap but have local copies of
bootstrap2 and 3 along with
any other resources they need.  everything is linked to files they provide.


do you know of another flask extension that utilizes bootstrap?  I think
ease of integration is key to usage?  therefore this PR could increase
adoption.

On Sat, Oct 1, 2016 at 7:51 AM, Goutham Balaraman <notifications@github.com>
wrote:

> @slippers <https://github.com/slippers> Just pinging to see if you have
> been using the flask_bootstrap in this project. Do you find the usage
> stable enough?
>
> I am still waiting to understand if this PR would have a wider use case
> for other users or not. Would help if more people would share the need
> though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gouthambs/Flask-Blogging/pull/73#issuecomment-250916397>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABy59a2ScIrDeZ5RnK2K7kSG4tMKkK5aks5qvnOOgaJpZM4JIOM->
> .
>

gouthambs commented:
I don't know of many that use flask-bootstrap. What I am pondering is, having flask bootstrap will require some one to understand the inner workings of flask-bootstrap. What if the main site doesn't use flask-bootstrap, then this PR in some sense forcing flask bootstrap on their project. I will look into flask-admin. Perhaps using local copies of bootstrap2 and 3 can be a way of extending to bootstrap3 i guess. Now aren't they out with version 4 as well?

slippers commented:
Hello,

Not sure either who else is using flask-bootstrap.

flask-bootstrap is loaded as a blueprint then it should not conflict with
other uses of pure bootstrap stuff.  unless they are using a different
version of flask-bootstrap.

I like flask-bootstrap for its configuration based features and helper
macros.  For instance in flask-blogging there was the google analytic
feature.  That is part of flask-bootstrap so there is one thing that
doesn't need to be maintained.  The ability to define where the resources
come from in the configuration is a great thing.

If you want to shelf this PR that is OK.  This seems to be raising many
reasonable questions.

Side thing: Wanted to know more about sqlalchemy so I built a variant of
storage using models
<https://github.com/slippers/Flask-Blogging/tree/models/flask_blogging/sqlastoragemodel>.
Passes the majority of the tests but missing some features like dynamic
table naming.  My curiosity is satisfied:)

On Tue, Oct 4, 2016 at 7:56 AM, Goutham Balaraman <notifications@github.com>
wrote:

> I don't know of many that use flask-bootstrap. What I am pondering is,
> having flask bootstrap will require some one to understand the inner
> workings of flask-bootstrap. What if the main site doesn't use
> flask-bootstrap, then this PR in some sense forcing flask bootstrap on
> their project. I will look into flask-admin. Perhaps using local copies of
> bootstrap2 and 3 can be a way of extending to bootstrap3 i guess. Now
> aren't they out with version 4 as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gouthambs/Flask-Blogging/pull/73#issuecomment-251413097>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABy59U0n4xcimdQNtUs63FM2uKMcbrztks5qwmk1gaJpZM4JIOM->
> .
>
